File: examples/code/guides/tool-callling/get-tools-helper.py
from typing import Optional
from arcadepy import NOT_GIVEN, AsyncArcade
from arcadepy.types import ToolDefinition


async def get_tool_definitions(
    client: Optional[AsyncArcade] = None,
    tools: Optional[list[str]] = None,
    toolkits: Optional[list[str]] = None,
    raise_on_empty: bool = True,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
) -> list[ToolDefinition]:
    """
    Retrieve tool definitions asynchronously from the Arcade client, accounting for pagination.

    Args:
        tools: Optional list of specific tool names to include.
        toolkits: Optional list of toolkit names to include all tools from.
        raise_on_empty: Whether to raise an error if no tools or toolkits are provided.
        limit: Optional limit on the number of tools to retrieve per request.
        offset: Optional offset for paginated requests.

    Returns:
        List of ToolDefinition instances.

    Raises:
        ValueError: If no tools or toolkits are provided and raise_on_empty is True.
    """
    if client is None:
        client = AsyncArcade()
    all_tools: list[ToolDefinition] = []

    # If no specific tools or toolkits are requested, raise an error.
    if not tools and not toolkits:
        if raise_on_empty:
            raise ValueError("No tools or toolkits provided to retrieve tool definitions.")
        return []

    # First, gather single tools if the user specifically requested them.
    if tools:
        for tool_id in tools:
            # ToolsResource.get(...) returns a single ToolDefinition.
            single_tool = await client.tools.get(name=tool_id)
            all_tools.append(single_tool)

    # Next, gather tool definitions from any requested toolkits.
    if toolkits:
        for tk in toolkits:
            paginated_tools = await client.tools.list(
                toolkit=tk,
                limit=NOT_GIVEN if limit is None else limit,
                offset=NOT_GIVEN if offset is None else offset,
            )
            async for tool in paginated_tools:
                all_tools.append(tool)

    return all_tools

File: examples/code/guides/tool-callling/get-tools-helper.ts
import Arcade from "@arcadeai/arcadejs";

export type ToolDefinition = Record<string, unknown>;

export type ArcadeClient = InstanceType<typeof Arcade>;

export type GetToolDefinitionsParams = {
  client?: ArcadeClient;
  tools?: string[];
  toolkits?: string[];
  raiseOnEmpty: boolean;
  limit?: number;
  offset?: number;
};
/**
 * Retrieve tool definitions from the Arcade client, accounting for pagination.
 *
 * - If `tools` are provided, fetch each by name via `client.tools.get`.
 * - If `toolkits` are provided, fetch lists via `client.tools.list` and collect `items`.
 * - If neither `tools` nor `toolkits` are provided and `raiseOnEmpty` is true, throw an error.
 */
export async function getToolDefinitions(
  params: GetToolDefinitionsParams,
): Promise<ToolDefinition[]> {
  const {
    client,
    tools,
    toolkits,
    raiseOnEmpty = true,
    limit,
    offset,
  } = params;
  const arcade = client ?? new Arcade();
  const allTools: ToolDefinition[] = [];

  const noTools = !tools || tools.length === 0;
  const noToolkits = !toolkits || toolkits.length === 0;

  if (noTools && noToolkits) {
    if (raiseOnEmpty) {
      throw new Error(
        "No tools or toolkits provided to retrieve tool definitions.",
      );
    }
    return [];
  }

  // Fetch specific tools by name
  if (tools && tools.length > 0) {
    const fetched = await Promise.all(
      tools.map((toolName) => arcade.tools.get(toolName)),
    );
    allTools.push(...fetched);
  }

  // Fetch tools from provided toolkits (respecting optional pagination)
  if (toolkits && toolkits.length > 0) {
    for (const tk of toolkits) {
      const response: any = await arcade.tools.list({
        toolkit: tk,
        ...(limit !== undefined ? { limit } : {}),
        ...(offset !== undefined ? { offset } : {}),
      });

      const items: ToolDefinition[] = Array.isArray(response?.items)
        ? response.items
        : [];
      allTools.push(...items);
    }
  }

  return allTools;
}

export default getToolDefinitions;

getToolDefinitions({
  client: new Arcade(),
});
